### PR TITLE
support psi

### PR DESCRIPTION
**Pull request by zouyee**
fix: https://github.com/containerd/containerd/issues/9411

### CODE REVIEW

**Comment by zouyee** [comment]
@dmcgowan PTAL

**Comment by samuelkarp** [comment]
Can you rebase this on the tip of `main` and remove the merge commit in the middle?

**Comment by zouyee** [comment]
> Can you rebase this on the tip of `main` and remove the merge commit in the middle?

done

File: cgroup2/utils.go [hunk 1]
```diff
@@ -479,6 +479,73 @@ func getHugePageSizeFromFilenames(fileNames []string) ([]string, error) {
...
+			err = parsePSIData(parts[1:], pv)
+			if err != nil {
+				logrus.Errorf("unable to parse psi data: %v", err)
+				return nil
+			}
+		}
+	}
+
+	if err := sc.Err(); err != nil {
+		return nil
```
Comment (fuweid) [inline cgroup2/utils.go:512]: could you log error here?

File: cgroup2/utils.go [hunk 2]
```diff
@@ -479,6 +479,73 @@ func getHugePageSizeFromFilenames(fileNames []string) ([]string, error) {
...
+	}
+	return psistats
+}
+
+func parsePSIData(psi []string, data *stats.PSIData) error {
+	for _, f := range psi {
+		kv := strings.SplitN(f, "=", 2)
+		if len(kv) != 2 {
+			return fmt.Errorf("invalid psi data: %q", f)
```
Comment (fuweid) [inline cgroup2/utils.go:521]: Please align `psi` with `PSI`

File: cgroup2/utils_test.go [hunk 3]
```diff
@@ -41,6 +47,36 @@ func TestParseCgroupFromReader(t *testing.T) {
...
+
+	fakeCgroupDir := t.TempDir()
+	statPath := filepath.Join(fakeCgroupDir, "cpu.pressure")
+
+	if err := os.WriteFile(statPath, []byte(examplePSIData), 0o644); err != nil {
+		t.Fatal(err)
+	}
+
+	st := getStatPSIFromFile(filepath.Join(fakeCgroupDir, "cpu.pressure"))
+	if diff := cmp.Diff(st, stats.PSIStats{
```
Comment (fuweid) [inline cgroup2/utils_test.go:62]: What is the benefit of using `go-cmp`? What about using `assert.Equal`?

Comment (zouyee) [reply cgroup2/utils_test.go]: done

File: cgroup2/utils.go (hunk 1 above)
Comment (zouyee) [reply cgroup2/utils.go:513]: ok

File: cgroup2/utils.go (hunk 2 above)
Comment (zouyee) [reply cgroup2/utils.go:522]: done

**Comment by zouyee** [comment]
> Basically, it looks good. But, would you please squash into one commit and add two unit tests for Stats to cover PSI? Thanks

done

File: cgroup2/utils.go [hunk 4]
```diff
@@ -479,6 +479,74 @@ func getHugePageSizeFromFilenames(fileNames []string) ([]string, error) {
...
```
Comment (fuweid) [inline cgroup2/utils.go:522]: s/psi/PSI

**Review by fuweid — APPROVED** [review]
LGTM

**Review by fuweid — APPROVED** [review]
LGTM

**Comment by zouyee** [comment]
@samuelkarp ping

**Review by dmcgowan — APPROVED** [review]

**Pull request merged**